src/extras.jl

```diff
@@ -142,36 +152,70 @@ function cut(x::AbstractArray{T, N}, breaks::AbstractVector;
+        to = map(x -> sprint(show, x, context=:compact=>true), breaks[2:n])
         levs = Vector{String}(undef, n-1)
         for i in 1:n-2
-            levs[i] = labels(from[i], to[i], i, closed=false)
+            levs[i] = labels(from[i], to[i], i,
+                             leftclosed=breaks[i] != breaks[i+1], rightclosed=false)
         end
-        levs[end] = labels(from[end], to[end], n-1, closed=extend)
+        levs[end] = labels(from[end], to[end], n-1,
+                           leftclosed=breaks[end-1] != breaks[end], rightclosed=extend)
     else
         length(labels) == n-1 || throw(ArgumentError("labels must be of length $(n-1), but got length $(length(labels))"))
         # Levels must have element type String for type stability of the result
         levs::Vector{String} = copy(labels)
     end
+    if !allunique(levs)
+        if labels === default_formatter
+            throw(ArgumentError("all labels must be unique, but `breaks` contains duplicates: " *
+                                "specify custom `labels` with unique names"))
+        else
+            throw(ArgumentError("all labels must be unique"))
+        end
+    end
 
     pool = CategoricalPool(levs, true)
     S = T >: Missing ? Union{String, Missing} : String
     CategoricalArray{S, N}(refs, pool)
 end
 
+"""
+    quantile_formatter(from, to, i; leftclosed, rightclosed)
+
+Provide the default label format for the `cut(x, ngroups)` method.
+"""
+quantile_formatter(from, to, i; leftclosed, rightclosed) =
+    string("Q", i, ": ", leftclosed ? "[" : "(", from, ", ", to, rightclosed ? "]" : ")")
+
 """
     cut(x::AbstractArray, ngroups::Integer;
-        labels::Union{AbstractVector{<:AbstractString},Function})
+        labels::Union{AbstractVector{<:AbstractString},Function},
+        allowempty::Bool=false)
 
 Cut a numeric array into `ngroups` quantiles, determined using
 [`quantile`](@ref).
+
+# Keyword arguments
+* `labels::Union{AbstractVector,Function}: a vector of strings giving the names to use for
+  the intervals; or a function `f(from, to, i; leftclosed, rightclosed)` that generates
+  the labels from the left and right interval boundaries and the group index. Defaults to
+  `"Qi: [from, to)"` (or `"Qi: [from, to]"` for the rightmost interval if `extend == true`).
+* `allowempty::Bool=false`: when `false`, an error is raised if some breaks appear
+  multiple times, generating empty intervals; when `true`, duplicate breaks are allowed
+  and the intervals they generate are kept as unused levels
+  (but duplicate labels are not allowed).
 """
-cut(x::AbstractArray, ngroups::Integer;
-    labels::Union{AbstractVector{<:AbstractString},Function}=default_formatter) =
-    cut(x, Statistics.quantile(x, (1:ngroups-1)/ngroups); extend=true, labels=labels)
+function cut(x::AbstractArray, ngroups::Integer;
+             labels::Union{AbstractVector{<:AbstractString},Function}=default_formatter,
```

Comment:
why it is `default_formatter` here and above you defined `quantile_formatter`?

Reply:
That's a typo.